Multi GPU Klein9B training early OOM

Training with release 4.0.6 a Klein9B on 2x 5090, it is failing early with OOM:
```
Fetching 4 files:  25%|██▌       | 1/4 [10:46<32:19, 646.51s/it]Download complete. Moving file to /workspace/huggingface/hub/models--black-forest-labs--FLUX.2-klein-base-9B/blobs/d58533b468c31caa0222540a8aefddea1d74dc6e1fee92
8da8556d3d85729d6e                                                                                                                                                                                                               
[RANK 0] 2026-02-13 15:06:28,152 [INFO] Download complete. Moving file to /workspace/huggingface/hub/models--black-forest-labs--FLUX.2-klein-base-9B/blobs/d58533b468c31caa0222540a8aefddea1d74dc6e1fee928da8556d3d85729d6e      
                                                                                                                                                                                                                                 
Fetching 4 files: 100%|██████████| 4/4 [11:22<00:00, 170.57s/it]                                                
                                                                                                                                                                                                                                 
Fetching 4 files: 100%|██████████| 4/4 [11:22<00:00, 170.56s/it]                                                                                                                                                                 
The following TP rules were not applied on any of the layers: {'layers.*.self_attn.q_proj': 'colwise', 'layers.*.self_attn.k_proj': 'colwise', 'layers.*.self_attn.v_proj': 'colwise', 'layers.*.self_attn.o_proj': 'rowwise', 'l
ayers.*.mlp.gate_proj': 'colwise', 'layers.*.mlp.up_proj': 'colwise', 'layers.*.mlp.down_proj': 'rowwise'}                                                                                                                       
The following layers were not sharded: model.layers.*.self_attn.k_norm.weight, model.layers.*.self_attn.q_norm.weight, lm_head.weight, model.norm.weight, model.layers.*.post_attention_layernorm.weight, model.embed_tokens.weig
ht, model.layers.*.input_layernorm.weight                                                                       
                                                                                                                                                                                                                                 
Loading checkpoint shards: 100%|██████████| 4/4 [00:00<00:00, 49.57it/s]                                                                                                                                                         
                                                                                                                                                                                                                                 
Could not locate the custom_generate/generate.py inside black-forest-labs/FLUX.2-klein-base-9B.                                                                                                                                  
[RANK 0] 2026-02-13 15:06:31,805 [INFO] Moving Qwen3 text encoder to cuda:0 (accelerator.device=cuda:0, local_rank=0, process_index=0)                                                                                           
[RANK 1] 2026-02-13 15:06:40,831 [ERROR] GPU circuit breaker OPEN: cuda_error - CUDA out of memory. Tried to allocate 1.16 GiB. GPU 0 has a total capacity of 31.36 GiB of which 1.83 GiB is free. Including non-PyTorch memory, 
this process has 14.76 GiB memory in use. Process 1351 has 14.76 GiB memory in use. Of the allocated memory 14.25 GiB is allocated by PyTorch, and 16.08 MiB is reserved by PyTorch but unallocated. If reserved but unallocated 
memory is large try setting PYTORCH_ALLOC_CONF=expandable_segments:True to avoid fragmentation.  See documentation for Memory Management  (https://pytorch.org/docs/stable/notes/cuda.html#environment-variables) (GPU 0)        
[RANK 0] 2026-02-13 15:06:40,832 [ERROR] GPU circuit breaker OPEN: cuda_error - CUDA out of memory. Tried to allocate 1.16 GiB. GPU 0 has a total capacity of 31.36 GiB of which 1.77 GiB is free. Process 1352 has 14.76 GiB mem
ory in use. Including non-PyTorch memory, this process has 14.76 GiB memory in use. Of the allocated memory 14.25 GiB is allocated by PyTorch, and 16.08 MiB is reserved by PyTorch but unallocated. If reserved but unallocated 
memory is large try setting PYTORCH_ALLOC_CONF=expandable_segments:True to avoid fragmentation.  See documentation for Memory Management  (https://pytorch.org/docs/stable/notes/cuda.html#environment-variables) (GPU 0)        
[RANK 0] 2026-02-13 15:06:40,832 [INFO] GPU fault webhook sent successfully                                                                                                                                                      
[RANK 0] 2026-02-13 15:06:40,860 [INFO] Broadcasting error event for job 2a6745e5: Training failed: OutOfMemoryError                                                                                                             
[RANK 0] 2026-02-13 15:06:40,861 [INFO] _update_job_store_status called with job_id=2a6745e5, status=failed                                                                                                                      
[RANK 0] 2026-02-13 15:06:40,861 [INFO] Scheduling _do_updates task for job 2a6745e5                                                                                                                                             
```

From the container start (done internally by vast), I see that the container gets passed:
```
["NVIDIA_VISIBLE_DEVICES", "2,3"], ["NV_GPU", "2,3"], ["GPU_COUNT", "2"],
```
but SSHing into the container I see:
```
$ printenv | grep VIS
NVIDIA_VISIBLE_DEVICES=all
```

See also #2583

--- a/simpletuner/helpers/training/trainer.py
+++ b/simpletuner/helpers/training/trainer.py
@@ -6596,6 +6596,16 @@ def run_trainer_job(config):
             if tokens:
                 selected_device_ids = tokens
 
+    def _normalize_visible_device_list(value: Optional[str]) -> Optional[str]:
+        if not isinstance(value, str):
+            return None
+        tokens = [token.strip() for token in value.split(",") if token.strip()]
+        if not tokens:
+            return None
+        if len(tokens) == 1 and tokens[0].lower() in {"all", "none", "void"}:
+            return None
+        return ",".join(tokens)
+
     def _resolve_hf_token() -> Optional[str]:
         possible_env_vars = (
             "HUGGINGFACEHUB_API_TOKEN",
@@ -6847,6 +6857,12 @@ def run_trainer_job(config):
 
         if selected_device_ids:
             launch_env["CUDA_VISIBLE_DEVICES"] = ",".join(selected_device_ids)
+        elif not _normalize_visible_device_list(launch_env.get("CUDA_VISIBLE_DEVICES")):
+            provider_visible_devices = _normalize_visible_device_list(launch_env.get("NV_GPU"))
+            if provider_visible_devices is None:
+                provider_visible_devices = _normalize_visible_device_list(launch_env.get("NVIDIA_VISIBLE_DEVICES"))
+            if provider_visible_devices is not None:
+                launch_env["CUDA_VISIBLE_DEVICES"] = provider_visible_devices
 
         config_backend_value: Optional[str] = None
         config_env_value: Optional[str] = None
@@ -6937,11 +6953,36 @@ def run_trainer_job(config):
         if signal_file_path:
             launch_env["SIMPLETUNER_ACCELERATE_SIGNAL_FILE"] = str(signal_file_path)
 
+        extra_args = []
+        if accelerate_extra_args:
+            try:
+                extra_args = shlex.split(str(accelerate_extra_args))
+            except ValueError:
+                launch_logger.warning("Failed to parse accelerate extra args; using raw string")
+                extra_args = [str(accelerate_extra_args)]
+
+        def _has_accelerate_distributed_selector(args: list[str]) -> bool:
+            selectors = {
+                "--multi_gpu",
+                "--cpu",
+                "--tpu",
+                "--use_deepspeed",
+                "--use_fsdp",
+                "--use_megatron_lm",
+            }
+            for arg in args:
+                option = arg.split("=", 1)[0]
+                if option in selectors:
+                    return True
+            return False
+
         cmd = ["accelerate", "launch"]
 
         if config_file_arg:
             cmd.append(f"--config_file={config_file_arg}")
         else:
+            if proc_count > 1 and not _has_accelerate_distributed_selector(extra_args):
+                cmd.append("--multi_gpu")
             cmd.extend(
                 [
                     f"--mixed_precision={launch_env.get('MIXED_PRECISION', 'bf16')}",
@@ -6957,13 +6998,6 @@ def run_trainer_job(config):
                 if same_network_value:
                     cmd.append("--same_network")
 
-        extra_args = []
-        if accelerate_extra_args:
-            try:
-                extra_args = shlex.split(str(accelerate_extra_args))
-            except ValueError:
-                launch_logger.warning("Failed to parse accelerate extra args; using raw string")
-                extra_args = [str(accelerate_extra_args)]
         if extra_args:
             cmd.extend(extra_args)
 

--- a/tests/test_trainer.py
+++ b/tests/test_trainer.py
@@ -3,6 +3,7 @@
 import importlib.machinery as machinery
 import importlib.util as _importlib_util
 import json
+import os
 import shutil
 import sys
 import tempfile
@@ -846,7 +847,7 @@ class TestTrainer(unittest.TestCase):
         self.assertTrue(instance.abort_called)
         self.assertTrue(instance.should_abort)
 
-    def test_run_trainer_job_honours_single_device_selection(self):
+    def _run_trainer_job_with_captured_popen(self, payload):
         from simpletuner.helpers.training import trainer as trainer_module
 
         captured = {}
@@ -882,18 +883,57 @@ class TestTrainer(unittest.TestCase):
             return DummyProcess()
 
         with patch("subprocess.Popen", side_effect=fake_popen):
-            result = trainer_module.run_trainer_job(
-                {
-                    "accelerate_visible_devices": [1],
-                    "--num_processes": 1,
-                }
-            )
+            result = trainer_module.run_trainer_job(payload)
 
-            self.assertEqual(result, 0)
+        return result, captured
+
+    def test_run_trainer_job_honours_single_device_selection(self):
+        result, captured = self._run_trainer_job_with_captured_popen(
+            {
+                "accelerate_visible_devices": [1],
+                "--num_processes": 1,
+            }
+        )
+
+        self.assertEqual(result, 0)
         self.assertIn("cmd", captured)
         self.assertEqual(captured["cmd"][0], "accelerate")
         self.assertEqual(captured["env"].get("CUDA_VISIBLE_DEVICES"), "1")
         self.assertFalse(any("--accelerate_visible_devices" in arg for arg in captured["cmd"]))
+
+    def test_run_trainer_job_enables_multi_gpu_for_multiple_processes(self):
+        result, captured = self._run_trainer_job_with_captured_popen(
+            {
+                "accelerate_visible_devices": [2, 3],
+                "--num_processes": 2,
+            }
+        )
+
+        self.assertEqual(result, 0)
+        self.assertIn("--multi_gpu", captured["cmd"])
+        self.assertIn("--num_processes=2", captured["cmd"])
+        self.assertEqual(captured["env"].get("CUDA_VISIBLE_DEVICES"), "2,3")
+
+    def test_run_trainer_job_uses_provider_gpu_assignment_when_unset(self):
+        with patch.dict(os.environ, {"NV_GPU": "2,3", "NVIDIA_VISIBLE_DEVICES": "all"}, clear=False):
+            os.environ.pop("CUDA_VISIBLE_DEVICES", None)
+            result, captured = self._run_trainer_job_with_captured_popen({"--num_processes": 2})
+
+        self.assertEqual(result, 0)
+        self.assertIn("--multi_gpu", captured["cmd"])
+        self.assertEqual(captured["env"].get("CUDA_VISIBLE_DEVICES"), "2,3")
+
+    def test_run_trainer_job_does_not_duplicate_accelerate_selector(self):
+        result, captured = self._run_trainer_job_with_captured_popen(
+            {
+                "--num_processes": 2,
+                "accelerate_extra_args": "--use_fsdp",
+            }
+        )
+
+        self.assertEqual(result, 0)
+        self.assertNotIn("--multi_gpu", captured["cmd"])
+        self.assertIn("--use_fsdp", captured["cmd"])
 
     def test_accelerate_manual_triggers_are_relayed(self):
         from simpletuner.helpers.training import trainer as trainer_module

